library/modules/Gui.cpp

```diff
@@ -182,23 +182,23 @@ DEFINE_GET_FOCUS_STRING_HANDLER(dwarfmode)
     if (game->main_interface.info.open) {
         newFocusString = baseFocus;
         newFocusString += "/Info";
-        newFocusString += "/" + enum_item_key(game->main_interface.info.current_mode);
+        newFocusString += '/' + enum_item_key(game->main_interface.info.current_mode);
```

Comment:
Are you sure these changes are correct? I would interpret `char` + anything as integer math, not string concatenation.

Reply:
Tested and it works out. std::string has an overload for this, just like +=

Reply:
Ok, looks like there has been an `operator+(char, std::string)` available since C++11: https://en.cppreference.com/w/cpp/string/basic_string/operator%2B

Wondering why the changes, though, since they did catch me off guard. They were bundled in a protobuf commit but looked unrelated to protobuf to me.

Reply:
Otherwise we fail on -Wrestrict. It's "related" to the protobuf change since we lost the absorbed -Wnorestrict

Reply:
...how the heck does `"a" + std::string("b")` trigger `-Wrestrict`? I realize that's calling `operator+()` under the hood, but that doesn't have `restrict` qualifiers on either of its arguments.

Reply:
It is truly one of the more esoteric error messages I've seen. I'm not at the computer right now, but the issue was that it constructs a temporary string with a corrupt (? Maybe just ill -defined?) buffer size

Reply:
No luck reproducing with
```cpp
#include <string>

template<typename T>
inline std::string enum_item_key(T val) {
    return "foo";
}

int main() {
    std::string s = "a";
    s += "b" + enum_item_key(3);
}
```

I'll see if I can reproduce with the full codebase.

Reply:
It only seems to happen for `"/"` literals inside the `switch` block. Outside, either `"/"` or `'/'` is fine. Weird.

```
In static member function ‘static constexpr std::char_traits<char>::char_type* std::char_traits<char>::copy(char_type*, const char_type*, std::size_t)’,
    inlined from ‘static constexpr void std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_S_copy(_CharT*, const _CharT*, size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’ at /usr/include/c++/12/bits/basic_string.h:423:21,
    inlined from ‘constexpr std::__cxx11::basic_string<_CharT, _Traits, _Allocator>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_M_replace(size_type, size_type, const _CharT*, size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’ at /usr/include/c++/12/bits/basic_string.tcc:532:22,
    inlined from ‘constexpr std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::replace(size_type, size_type, const _CharT*, size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’ at /usr/include/c++/12/bits/basic_string.h:2171:19,
    inlined from ‘constexpr std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::insert(size_type, const _CharT*) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’ at /usr/include/c++/12/bits/basic_string.h:1928:22,
    inlined from ‘constexpr std::__cxx11::basic_string<_CharT, _Traits, _Allocator> std::operator+(const _CharT*, __cxx11::basic_string<_CharT, _Traits, _Allocator>&&) [with _CharT = char; _Traits = char_traits<char>; _Alloc = allocator<char>]’ at /usr/include/c++/12/bits/basic_string.h:3541:36,
    inlined from ‘bool DFHack::Gui::autoDFAnnouncement(df::report_init, std::string)’ at dfhack/library/modules/Gui.cpp:1692:37:
/usr/include/c++/12/bits/char_traits.h:431:56: error: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ accessing 9223372036854775810 or more bytes at offsets [2, 9223372036854775807] and 1 may overlap up to 9223372036854775813 bytes at offset -3 [-Werror=restrict]
  431 |         return static_cast<char_type*>(__builtin_memcpy(__s1, __s2, __n));
      |                                        ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
```

Reply:
Preliminary searching around suggests that this is a case of https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104336 or https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105329.

Someone else ran into the same issue in https://github.com/DFHack/dfhack/issues/2250#issuecomment-1196549459.

I guess this only happens for me (and you?) in C++20 with GCC 12? I wasn't able to get it to fail earlier, and curiously, CI didn't fail before your change either. (It is running GCC 12.1, vs 12.2 for me locally)

Reply:
The lack of CI failure may just be because I didn't push until it built successfully locally. This being a bug makes sense since the code we had looked just fine.